File: integrations/adk/typescript/src/memory-service.ts
/**
 * `ZepMemoryService` — an ADK-native `BaseMemoryService` backed by Zep.
 *
 * ADK's memory extension point (`BaseMemoryService`) lets a `Runner` search
 * long-term memory through a model-opt-in tool (`loadMemory` /
 * `preloadMemory`): the model decides when to call
 * `toolContext.searchMemory(query)`, which reaches this service via
 * `new Runner({ ..., memoryService: new ZepMemoryService(...) })`.
 *
 * This is a different extension point from `ZepContextTool` /
 * `createZepBeforeModelCallback`:
 *
 * - **ZepMemoryService** (this module) — ADK-native, model-opt-in. The model
 *   decides, per turn, whether to call `loadMemory`/`preloadMemory`. Use this
 *   when you want the model to actively decide when memory is relevant, or
 *   when wiring into ADK code paths that already expect a `memoryService`
 *   (e.g. ADK's own memory tools, evaluation harnesses).
 * - **ZepContextTool** / **createZepBeforeModelCallback** — guaranteed
 *   injection. Runs on every LLM turn, so Zep context is always present
 *   regardless of whether the model would have thought to ask for it. Use
 *   this for the common case: an assistant that should always have the
 *   user's long-term context available.
 *
 * The two are complementary, not mutually exclusive — e.g. use
 * `ZepContextTool` for always-on context and additionally register
 * `ZepMemoryService` so the model can explicitly search for more via
 * `loadMemory` when it decides the always-on context wasn't enough.
 *
 * Why `addSessionToMemory` is a no-op: Zep persistence already happens live,
 * on every turn, via `ZepContextTool` (or `createZepBeforeModelCallback` /
 * `createZepAfterModelCallback`, which call `thread.addMessages` directly).
 * ADK calls `addSessionToMemory` to flush a session's conversation into a
 * memory store at some point after it happens (e.g. session end); since Zep
 * already has every message as it's persisted turn-by-turn, doing it again
 * here would re-ingest the same conversation into the graph a second time.
 * This mirrors the Go integration's `NewMemoryService` and Python's
 * `ZepMemoryService`, whose `AddSessionToMemory`/`add_session_to_memory` are
 * the same intentional no-op for the same reason.
 */

import type { ZepClient } from "@getzep/zep-cloud";
import type {
  BaseMemoryService,
  MemoryEntry,
  SearchMemoryRequest,
  SearchMemoryResponse,
} from "@google/adk";
import type { Content } from "@google/genai";
import {
  searchGraphScope,
  type ZepGraphSearchScope,
} from "./graph-search-tool.js";
import { defaultLogger, type Logger } from "./logging.js";

// Minimal structural type for the `Session` ADK passes to
// `addSessionToMemory`. The method is a no-op (see module docstring), so the
// integration never needs to read a session's fields.
interface SessionLike {
  id?: string;
}

/**
 * Scopes supported by {@link ZepMemoryService.searchMemory}. Matches the
 * scope enum exposed by `ZepGraphSearchTool`.
 */
const SUPPORTED_SCOPES: readonly ZepGraphSearchScope[] = [
  "edges",
  "nodes",
  "episodes",
  "observations",
  "thread_summaries",
  "auto",
];

const DEFAULT_SCOPE: ZepGraphSearchScope = "edges";

/** Options for the {@link ZepMemoryService} constructor. */
export interface ZepMemoryServiceOptions {
  /** An initialised `ZepClient`. The caller owns its lifecycle. */
  zep: ZepClient;
  /**
   * Fixed graph UUID. When set, every `searchMemory` call targets this
   * graph. When omitted, the service reads the graph UUID of the user named
   * by `SearchMemoryRequest.userId`, which ADK carries from the session. In
   * Zep v4 that value must be the UUID of the Zep user.
   */
  graphUuid?: string;
  /**
   * The Zep graph search scope to use for every `searchMemory` call.
   * Matches `ZepGraphSearchTool`'s scope enum: `"edges"` (facts, the
   * default), `"nodes"` (entities and summaries), `"episodes"` (raw
   * message/data content), `"observations"` (derived memories),
   * `"thread_summaries"` (incremental thread summaries), or `"auto"` (Zep's
   * own pre-assembled Context Block, returned as a single memory entry).
   */
  scope?: ZepGraphSearchScope;
  /**
   * Maximum number of results per search. `undefined` (the default) omits
   * the parameter so the Zep SDK applies its own default.
   */
  limit?: number;
  /** Logger for Zep failures and unsupported-scope warnings. Defaults to a `console`-backed logger. */
  logger?: Logger;
}

/**
 * ADK-native memory service that searches a user's Zep knowledge graph.
 *
 * Register this on a `Runner` to give ADK's built-in memory tooling
 * (`loadMemory`, `preloadMemory`) access to Zep:
 *
 * @example
 * ```ts
 * import { Runner, LlmAgent } from "@google/adk";
 * import { LOAD_MEMORY } from "@google/adk";
 * import { ZepClient } from "@getzep/zep-cloud";
 * import { ZepMemoryService } from "@getzep/zep-adk";
 *
 * const zep = new ZepClient({ apiKey: process.env.ZEP_API_KEY! });
 * const agent = new LlmAgent({
 *   name: "memory_agent",
 *   model: "gemini-2.5-flash",
 *   tools: [LOAD_MEMORY],
 * });
 * const runner = new Runner({
 *   agent,
 *   appName: "my_app",
 *   sessionService,
 *   memoryService: new ZepMemoryService({ zep }),
 * });
 * ```
 *
 * See the module docstring for guidance on when to use this versus
 * `ZepContextTool` / `createZepBeforeModelCallback`.
 */
export class ZepMemoryService implements BaseMemoryService {
  private readonly zep: ZepClient;
  private readonly graphUuid?: string;
  private readonly scope: ZepGraphSearchScope;
  private readonly limit?: number;
  private readonly logger: Logger;
  /** Graph UUID for each user UUID, cached for the life of the service. */
  private readonly graphUuidByUser = new Map<string, string>();

  constructor(options: ZepMemoryServiceOptions) {
    this.zep = options.zep;
    this.graphUuid = options.graphUuid;
    this.scope = options.scope ?? DEFAULT_SCOPE;
    this.limit = options.limit;
    this.logger = options.logger ?? defaultLogger;
  }

  /**
   * No-op. See the module docstring for the no-double-persist rationale.
   *
   * Conversation turns are already persisted live via `ZepContextTool` (or
   * `createZepBeforeModelCallback`/`createZepAfterModelCallback`). Ingesting
   * the session again here would duplicate that work in Zep's graph.
   */
  async addSessionToMemory(_session: SessionLike): Promise<void> {
    this.logger.debug(
      "ZepMemoryService.addSessionToMemory is a no-op: Zep already ingests " +
        "conversation turns live via ZepContextTool / " +
        "createZepBeforeModelCallback + createZepAfterModelCallback, so " +
        "re-ingesting the full session here would double-persist it.",
    );
  }

  /**
   * Search the user's Zep graph and map results to `MemoryEntry` objects.
   *
   * `request.appName` has no Zep equivalent — Zep scopes memory by user
   * graph, not by application — so it is accepted (to satisfy the ADK
   * interface) but not forwarded to Zep.
   *
   * Zep v4 searches a graph by UUID. The service uses the configured
   * `graphUuid` when there is one. Otherwise it reads the graph UUID of
   * `request.userId` with `user.get`, which requires the ADK session to
   * carry the UUID of the Zep user. The service never calls `user.lookup`.
   *
   * On any Zep failure, logs a warning (lengths only, never query or result
   * content) and returns an empty response rather than throwing, so a
   * memory lookup never breaks the agent.
   *
   * An unsupported scope is rejected before the search is issued: like Go's
   * `memoryService.SearchMemory` (see `searchScopeSupported` in
   * `integrations/adk/go/memory.go`) and Python's `ZepMemoryService`, this
   * avoids spending a live network call on a scope we can never map into
   * memory entries.
   */
  async searchMemory(
    request: SearchMemoryRequest,
  ): Promise<SearchMemoryResponse> {
    if (!SUPPORTED_SCOPES.includes(this.scope)) {
      this.logger.warn(
        `Unsupported Zep memory search scope '${this.scope}'; returning no memories`,
      );
      return { memories: [] };
    }

    let texts: string[];
    try {
      const graphUuid = await this.resolveGraphUuid(request.userId);
      texts = await searchGraphScope(this.zep, graphUuid, {
        scope: this.scope,
        body: { query: request.query },
        limit: this.limit,
      });
    } catch (error) {
      this.logger.warn(
        `Zep memory search failed (userId_len=${request.userId.length}, query_len=${request.query.length})`,
        error,
      );
      return { memories: [] };
    }

    const memories: MemoryEntry[] = texts.map((text) => ({
      content: { role: "model", parts: [{ text }] } satisfies Content,
      author: "Zep",
    }));
    return { memories };
  }

  /** Resolve the UUID of the graph to search for a user UUID. */
  private async resolveGraphUuid(userUuid: string): Promise<string> {
    if (this.graphUuid) {
      return this.graphUuid;
    }

    const cached = this.graphUuidByUser.get(userUuid);
    if (cached) {
      return cached;
    }

    const user = await this.zep.user.get(userUuid);
    if (!user.graphUuid) {
      throw new Error(`Zep user ${userUuid} has no graph UUID.`);
    }
    this.graphUuidByUser.set(userUuid, user.graphUuid);
    return user.graphUuid;
  }
}

File: integrations/adk/python/src/zep_adk/memory_service.py
"""
ZepMemoryService -- an ADK-native ``BaseMemoryService`` backed by Zep.

ADK's memory extension point (``google.adk.memory.base_memory_service.BaseMemoryService``)
lets a ``Runner`` search long-term memory through a model-opt-in tool
(``load_memory`` / ``preload_memory``): the model decides when to call
``tool_context.search_memory(query)``, which reaches this service via
``Runner(..., memory_service=ZepMemoryService(...))``.

This is a different extension point from :class:`~zep_adk.context_tool.ZepContextTool`:

* **ZepMemoryService** (this module) -- ADK-native, model-opt-in. The model
  decides, per turn, whether to call ``load_memory``/``preload_memory``. Use
  this when you want the model to actively decide when memory is relevant,
  or when wiring into ADK code paths that already expect a
  ``memory_service`` (e.g. ADK's own memory tools, evaluation harnesses).
* **ZepContextTool** -- guaranteed injection. Runs on every LLM turn via
  ``process_llm_request``, so Zep context is always present regardless of
  whether the model would have thought to ask for it. Use this for the
  common case: an assistant that should always have the user's long-term
  context available.

The two are complementary, not mutually exclusive -- e.g. use
``ZepContextTool`` for always-on context and additionally register
``ZepMemoryService`` so the model can explicitly search for more via
``load_memory`` when it decides the always-on context wasn't enough.

Why ``add_session_to_memory`` is a no-op: Zep persistence already happens
live, on every turn, via :class:`~zep_adk.context_tool.ZepContextTool` (or a
custom integration using ``thread.add_messages``) and
:func:`~zep_adk.callbacks.create_after_model_callback`. ADK calls
``add_session_to_memory`` to flush a session's conversation into a memory
store at some point after it happens (e.g. session end); since Zep already
has every message as it's persisted turn-by-turn, doing it again here would
re-ingest the same conversation into the graph a second time. This mirrors
the Go integration's ``NewMemoryService``, whose ``AddSessionToMemory`` is
the same intentional no-op for the same reason.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from google.adk.memory.base_memory_service import BaseMemoryService, SearchMemoryResponse
from google.adk.memory.memory_entry import MemoryEntry
from google.genai import types
from typing_extensions import override
from zep_cloud.client import AsyncZep

from .graph_search_tool import scope_results_to_texts

if TYPE_CHECKING:
    from google.adk.sessions import Session

logger = logging.getLogger(__name__)

#: Scopes handled by treating ``result.context`` as a single pre-materialized
#: Context Block, rather than a list of discrete edges/nodes/etc.
_AUTO_SCOPE = "auto"

#: Scopes supported by :meth:`ZepMemoryService.search_memory`. Matches the
#: scope enum exposed by :class:`~zep_adk.graph_search_tool.ZepGraphSearchTool`.
_SUPPORTED_SCOPES = frozenset(
    {"edges", "nodes", "episodes", "observations", "thread_summaries", _AUTO_SCOPE}
)


class ZepMemoryService(BaseMemoryService):
    """ADK-native memory service that searches a user's Zep knowledge graph.

    Register this on a ``Runner`` to give ADK's built-in memory tooling
    (``load_memory``, ``preload_memory``) access to Zep:

    .. code-block:: python

        from google.adk.runners import Runner
        from google.adk.tools import load_memory
        from zep_adk import ZepMemoryService

        agent = Agent(
            name="my_agent",
            model="gemini-2.5-flash",
            tools=[load_memory],
        )
        runner = Runner(
            agent=agent,
            app_name="my_app",
            session_service=session_service,
            memory_service=ZepMemoryService(zep=zep),
        )

    See the module docstring for guidance on when to use this versus
    :class:`~zep_adk.context_tool.ZepContextTool`.

    Args:
        zep: An initialised ``AsyncZep`` client.
        scope: The Zep graph search scope to use for every ``search_memory``
            call. Matches :class:`~zep_adk.graph_search_tool.ZepGraphSearchTool`'s
            scope enum: ``"edges"`` (facts, the default), ``"nodes"``
            (entities and summaries), ``"episodes"`` (raw message/data
            content), ``"observations"`` (derived memories),
            ``"thread_summaries"`` (incremental thread summaries), or
            ``"auto"`` (Zep's own pre-assembled Context Block, returned as a
            single memory entry).
        limit: Maximum number of results per search. ``None`` (the default)
            omits the parameter so the Zep SDK applies its own default.
    """

    def __init__(
        self,
        *,
        zep: AsyncZep,
        scope: str = "edges",
        limit: int | None = None,
    ) -> None:
        self._zep: AsyncZep = zep
        self._scope: str = scope
        self._limit: int | None = limit

    @override
    async def add_session_to_memory(self, session: Session) -> None:
        """No-op. See the module docstring for the no-double-persist rationale.

        Conversation turns are already persisted live via
        :class:`~zep_adk.context_tool.ZepContextTool` (or an equivalent
        ``thread.add_messages`` call) and
        :func:`~zep_adk.callbacks.create_after_model_callback`. Ingesting the
        session again here would duplicate that work in Zep's graph.
        """
        logger.debug(
            "ZepMemoryService.add_session_to_memory is a no-op: Zep already "
            "ingests conversation turns live via ZepContextTool / "
            "create_after_model_callback, so re-ingesting the full session "
            "here would double-persist it."
        )

    @override
    async def search_memory(
        self,
        *,
        app_name: str,
        user_id: str,
        query: str,
    ) -> SearchMemoryResponse:
        """Search the user's Zep graph and map results to ``MemoryEntry`` objects.

        ``app_name`` has no Zep equivalent -- Zep scopes memory by user
        graph, not by application -- so it is accepted (to satisfy the ADK
        interface) but not forwarded to Zep.

        On any Zep failure, logs a warning (lengths/counts only, never query
        or result content) and returns an empty response rather than raising,
        so a memory lookup never breaks the agent.

        An unsupported scope is rejected before the search is issued: like
        Go's ``memoryService.SearchMemory`` (see ``searchScopeSupported`` in
        ``integrations/adk/go/memory.go``), this avoids spending a live
        network call on a scope we can never map into memory entries.
        """
        if self._scope not in _SUPPORTED_SCOPES:
            logger.warning(
                "Unsupported Zep memory search scope %r; returning no memories",
                self._scope,
            )
            return SearchMemoryResponse()

        search_kwargs: dict[str, Any] = {
            "user_id": user_id,
            "query": query,
            "scope": self._scope,
        }
        if self._limit is not None:
            search_kwargs["limit"] = self._limit

        try:
            result = await self._zep.graph.search(**search_kwargs)
        except Exception as exc:
            logger.warning(
                "Zep memory search failed (user_id_len=%d, query_len=%d): %s",
                len(user_id),
                len(query),
                exc,
            )
            return SearchMemoryResponse()

        texts = self._result_texts(result)
        memories = [
            MemoryEntry(
                content=types.Content(parts=[types.Part(text=text)], role="model"),
                author="Zep",
            )
            for text in texts
        ]
        return SearchMemoryResponse(memories=memories)

    def _result_texts(self, result: Any) -> list[str]:
        """Flatten a ``graph.search`` result into text items for the configured scope."""
        if self._scope == _AUTO_SCOPE:
            context: str | None = getattr(result, "context", None)
            if context and context.strip():
                return [context.strip()]
            return []

        return scope_results_to_texts(result, self._scope)
